src/library.js

```diff
@@ -51,6 +51,221 @@ LibraryManager.library = {
+    idx -= 1;
+    stream.inc_pos_before_read = 1;
+
+    var id;
+    var type;
+    var name = stream.getdents[stream.getdents.length - 1 - idx];
+    if (name === '.' || name == '..') {
+      id = 1;
+      type = 4; // DT_DIR
+    } else {
+      var child = FS.lookupNode(stream.node, name);
+      id = child.id;
+      type = FS.isChrdev(child.mode) ? 2 :  // DT_CHR, character device.
+      FS.isDir(child.mode) ? 4 :     // DT_DIR, directory.
+      FS.isLink(child.mode) ? 10 :   // DT_LNK, symbolic link.
+      8;                             // DT_REG, regular file.
+    }
+    {{{ makeSetValue('entry', C_STRUCTS.dirent.d_ino, 'id', 'i32') }}};
+    {{{ makeSetValue('entry', C_STRUCTS.dirent.d_off, 'idx + 1', 'i32') }}};
+    {{{ makeSetValue('entry', C_STRUCTS.dirent.d_reclen, C_STRUCTS.dirent.__size__, 'i16') }}};
+    {{{ makeSetValue('entry', C_STRUCTS.dirent.d_type, 'type', 'i8') }}};
+    stringToUTF8(name, entry +  {{{ C_STRUCTS.dirent.d_name }}}, 256);
+    {{{ makeSetValue('result', '0', 'entry', 'i8*') }}};
+    return 0;
+  },
+#if USE_PTHREADS
+  __readdir_result: '; if (ENVIRONMENT_IS_PTHREAD) ___readdir_result = PthreadWorkerInit.___readdir_result; else PthreadWorkerInit.___readdir_result = ___readdir_result = {{{ makeStaticAlloc(4) }}}',
+#else
+  __readdir_result: '{{{ makeStaticAlloc(4) }}}',
+#endif
+  readdir__deps: ['readdir_r', '__setErrNo', '$ERRNO_CODES', '__readdir_result'],
+  readdir: function(dirp) {
+    // struct dirent *readdir(DIR *dirp);
+    // http://pubs.opengroup.org/onlinepubs/007908799/xsh/readdir_r.html
+    var stream;
+    try {
+      stream = __getStreamFromPtr(dirp);
+    } catch (e) {
+      if (!(e instanceof FS.ErrnoError)) throw e + ' : ' + stackTrace();
+      return e.errno;
+    }
+    if (!stream.readdir_current) {
+      stream.readdir_current = _malloc({{{ C_STRUCTS.dirent.__size__ }}});
+      if (!stream.readdir_current) {
+        return ERRNO_CODES.ENOMEM;
+      }
+    }
+    var err = _readdir_r(dirp, stream.readdir_current, ___readdir_result);
+    if (err) {
+      ___setErrNo(err);
+      return 0;
+    }
+    return {{{ makeGetValue(0, '___readdir_result', 'i8*') }}};
+  },
+  dirfd__deps: ['$FS', '_getStreamFromPtr'],
+  dirfd: function(dirp) {
+    try {
+      return __getStreamFromPtr(dirp).fd;
+    } catch (e) {
+      FS.handleFSError(e);
+      return -1;
+    }
+  },
+  dirfd__deps: ['$FS', '_getStreamFromPtr'],
+  fdopendir: function(fd) {
+    try {
+      var stream = FS.getStream(fd);
+      if (!stream) {
+        throw new FS.ErrnoError(ERRNO_CODES.EBADF);
+      }
+
+      if (!FS.isDir(stream.node.mode)) {
+        throw new FS.ErrnoError(ERRNO_CODES.ENOTDIR);
+      }
+
+      // just read all ents upfront
+      stream.getdents = FS.readdir(stream.path);
+      stream.getdentsposition = 0;
+      stream.inc_pos_before_read = 0;
+
+      return stream.fd + 1;
+    } catch (e) {
+      FS.handleFSError(e);
+      return 0;
+    }
+  },
+
```

Comment:
Any reason all this can't live in library_fs.js?

Reply:
Hmm, I was under the impression that `library_fs.js` was an Emscripten-internal support library while `library.js` was the C library. I could move this functionality into `library_fs.js` but then I would have to duplicate some support code in other FS implementations (e.g. asmfs).

Reply:
I'm not an expert in this area.  I'll defer to @kripken

Reply:
I think the main question is, should this code be present when the user sets `-s FILESYSTEM=0`. If not, then `library_fs.js` is the right place.

(We should probably move some other `library` code into `library_fs`, btw.)

Reply:
I don't think these function should exist for users who set `-s FILESYSTEM=0`, so yes it sounds like they should be in the `_fs.js` file